### ISSUE TITLE
Netty causes ClassLoader leak in container environments

### ISSUE THREAD

**Issue by kevinherron** [netty/netty#6565]
I'm running Netty in a container-like environment where modules can be loaded and unloaded. Unloading a module that has used Netty causes the ClassLoader to be leaked.

I'm able to reproduce with this minimal example (running in the container environment, obviously):

```java
public class GatewayHook extends AbstractGatewayModuleHook {

    private final Logger logger = LoggerFactory.getLogger(getClass());

    private final NioEventLoopGroup eventLoop = new NioEventLoopGroup(0);

    @Override
    public void setup(GatewayContext gatewayContext) {
        logger.info("setup()");
    }

    @Override
    public void startup(LicenseState licenseState) {
        logger.info("startup()");

        try {
            Bootstrap bootstrap = new Bootstrap();

            bootstrap.group(eventLoop)
                .channel(NioSocketChannel.class)
                .option(ChannelOption.CONNECT_TIMEOUT_MILLIS, 5000)
                .option(ChannelOption.TCP_NODELAY, true)
                .handler(new ChannelInitializer<SocketChannel>() {
                    @Override
                    protected void initChannel(SocketChannel socketChannel) throws Exception {
                        logger.info("initChannel");
                    }
                });

            bootstrap.connect("localhost", 1234).get(2, TimeUnit.SECONDS);
        } catch (Throwable t) {
            logger.error("failed getting un-gettable endpoints: {}", t.getMessage(), t);
        }
    }

    @Override
    public void shutdown() {
        logger.info("shutdown()");

        try {
            eventLoop.shutdownGracefully().get();
        } catch (Throwable e) {
            logger.error("Error waiting for event loop shutdown: {}", e.getMessage(), e);
        }
        try {
            GlobalEventExecutor.INSTANCE.shutdownGracefully().get();
        } catch (Throwable e) {
            logger.error("Error waiting for GlobalEventExecutor shutdown: {}", e.getMessage(), e);
        }
        try {
            DefaultAddressResolverGroup.INSTANCE.close();
        } catch (Throwable e) {
            logger.error("Error closing DefaultAddressResolverGroup: {}", e.getMessage(), e);
        }
        InternalThreadLocalMap.destroy();
        FastThreadLocal.removeAll();
    }

}
```

I've tried shutting down as many global looking Netty resources as I could find but it still leaks. I've uploaded a JProfiler heap dump [here](https://www.dropbox.com/s/ylmfieul8oao49c/netty.jps?dl=0). If you group by ClassLoader and then select the only `NonLockingURLClassLoader` instance you'll see the Netty classes still hanging around.

I have of course tested this without invoking any Netty code and it does not leak.

### Netty version
Netty 4.1.9.Final

### JVM version (e.g. `java -version`)
java version "1.8.0_112"

### OS version (e.g. `uname -a`)
macOS 10.12.3

**Comment by normanmaurer** [comment]
@kevinherron can you try calling `FastThreadLocal.removeAll()` as last method before complete the shutdown method ?

**Comment by kevinherron** [comment]
@normanmaurer I added that and tested again, still leaks.

**Comment by normanmaurer** [comment]
@kevinherron can you provide me with something that I can easily run to reproduce ?

**Comment by normanmaurer** [comment]
@kevinherron also can you please share a standard dump that can be opened with an OSS tool ?

**Comment by kevinherron** [comment]
@normanmaurer [here's a dump in hprof format](https://www.dropbox.com/s/zaxzltkz6gtrwym/netty-leak.hprof?dl=0)

Can I email you instructions for setting up an environment to reproduce? It should take less than 5 minutes from downloading the software to reproducing.

**Comment by normanmaurer** [comment]
@kevinherron norman dot maurer at googlemail dot com ;)

**Comment by kevinherron** [comment]
@normanmaurer did you get my email?

**Comment by normanmaurer** [comment]
@kevinherron yes but had no time yet :(

**Comment by kevinherron** [comment]
@normanmaurer ok that's no problem; I just wanted to make sure it arrived.

**Comment by slandelle** [comment]
It's not (or not only) a ClassLoader leak.
The `io.netty.util.concurrent.GlobalEventExecutor` singleton is still up (along with a `FastThreadLocalThread` instance).

**Comment by kevinherron** [comment]
@slandelle right... that's what leads to the ClassLoader leaking, is that there's a handful of Netty bits still running or referenced that I can't figure out how to shut down.

**Comment by slandelle** [comment]
Could be Finalizers.
Could be some Netty resources being cached ThreadLocals while they are being loaded from a global thread that the unloaded module doesn't own.

**Comment by Scottmitch** [comment]
ideally we can provide a reasonably centralized place to cleanup/unload static state.

**Comment by normanmaurer** [comment]
@kevinherron is this issue still present ? I suspect I fixed it

**Comment by kevinherron** [comment]
@normanmaurer funny timing... I've got a backlog card for upgrading Netty to latest and testing using the proof of concept I developed. I'll let you know soon.

**Comment by kevinherron** [comment]
@normanmaurer is there a particular minimum Netty version you want me to use?

**Comment by johnou** [comment]
@kevinherron go for 4.1.22.

**Comment by kevinherron** [comment]
It doesn't appear to be fixed... still a bunch of Netty classes showing up in JProfiler after unloading the module.

[HPROF before unloading the module](https://www.dropbox.com/s/fodandm2xe3zmdy/before-unload.hprof?dl=0)
[HPROF after unloading the module and a GC run](https://www.dropbox.com/s/ipvnogfa7hwsp1j/after-unload.hprof?dl=0)

The only class provided by the PoC module is `com.kevinherron.GatewayHook`, which you can see is properly unloaded; it's just some Netty classes still hanging around. Everything else in the dump that isn't `io.netty` or `com.kevinherron` is part of the container.

**Comment by normanmaurer** [comment]
@kevinherron could you create a small reproducer and explain how to run it ?

**Comment by kevinherron** [comment]
@normanmaurer last year I emailed you instructions, but they may have been blocked because of the attachment.

I've updated the README file of the repro repo to include full instructions to run: https://github.com/kevinherron/netty-leak

**Comment by kevinherron** [comment]
@normanmaurer do you need anything else?

**Comment by normanmaurer** [comment]
I think we fixed this. Please upgrade and let us know if you still have the problem

**Comment by kevinherron** [comment]
> I think we fixed this. Please upgrade and let us know if you still have the problem 

In 4.1 or 4.2?

**Comment by normanmaurer** [comment]
both should have fixed it
